From patchwork Mon Feb  4 17:16:49 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Krish Sadhukhan <krish.sadhukhan@oracle.com>
X-Patchwork-Id: 10796203
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 7999A6C2
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  4 Feb 2019 17:41:57 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 694822BEE6
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  4 Feb 2019 17:41:57 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 667B22BEF3; Mon,  4 Feb 2019 17:41:57 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	UNPARSEABLE_RELAY autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 0E9982BEF0
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  4 Feb 2019 17:41:57 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728160AbfBDRlz (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 4 Feb 2019 12:41:55 -0500
Received: from userp2120.oracle.com ([156.151.31.85]:41992 "EHLO
        userp2120.oracle.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726585AbfBDRlz (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 4 Feb 2019 12:41:55 -0500
Received: from pps.filterd (userp2120.oracle.com [127.0.0.1])
        by userp2120.oracle.com (8.16.0.27/8.16.0.27) with SMTP id
 x14HXm7o025276;
        Mon, 4 Feb 2019 17:41:50 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : in-reply-to : references; s=corp-2018-07-02;
 bh=/OI7QCp2YemO24oc5Kj6DjMioLcdjbwDokQ66t1lNxU=;
 b=E7PdTV8OmEVp/6/hck+d9paEODXH+noz82v4c79ARFzxM13kNsFGjBkQpSnj/hBpQiuN
 RR9oZq+mQQg0HREtWrFi+t7eMG2eGW4JWkHuTydtOQGnIrKE6oi1kDQBLVnOqbHQrwhz
 NvN0c2K75oZr4NnEeFqKn/XQJbX0UrEAgR8bw7/kv+Ih9XBr3SoyjDZTUG/pSLyU4C7t
 ieyzcqmJWx+bB29wtjNBhgwXLNgxc3IbD5MpxohISm0/EQYzQdEczbeoWzUiZ9I+X3Zc
 B9YLT5vYSTltzrMwCgDY8Vp9kuzaMLnQu+QM584AnPnFQZ4VdyHzUy/w/uHtaopdLGQB tw==
Received: from aserv0021.oracle.com (aserv0021.oracle.com [141.146.126.233])
        by userp2120.oracle.com with ESMTP id 2qd98mxmfj-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Mon, 04 Feb 2019 17:41:49 +0000
Received: from userv0121.oracle.com (userv0121.oracle.com [156.151.31.72])
        by aserv0021.oracle.com (8.14.4/8.14.4) with ESMTP id x14HfmmG029144
        (version=TLSv1/SSLv3 cipher=DHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Mon, 4 Feb 2019 17:41:49 GMT
Received: from abhmp0018.oracle.com (abhmp0018.oracle.com [141.146.116.24])
        by userv0121.oracle.com (8.14.4/8.13.8) with ESMTP id x14HfmMu015749;
        Mon, 4 Feb 2019 17:41:48 GMT
Received: from ban25x6uut29.us.oracle.com (/10.153.73.29)
        by default (Oracle Beehive Gateway v4.0)
        with ESMTP ; Mon, 04 Feb 2019 17:41:47 +0000
From: Krish Sadhukhan <krish.sadhukhan@oracle.com>
To: kvm@vger.kernel.org
Cc: pbonzini@redhat.com, rkrcmar@redhat.com, jmattson@google.com
Subject: [PATCH 1/2][KVM nVMX]: Add a vmentry check for HOST_SYSENTER_ESP and
 HOST_SYSENTER_EIP fields
Date: Mon,  4 Feb 2019 12:16:49 -0500
Message-Id: <20190204171650.6920-2-krish.sadhukhan@oracle.com>
X-Mailer: git-send-email 2.17.2
In-Reply-To: <20190204171650.6920-1-krish.sadhukhan@oracle.com>
References: <20190204171650.6920-1-krish.sadhukhan@oracle.com>
X-Proofpoint-Virus-Version: vendor=nai engine=5900 definitions=9157
 signatures=668682
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0
 priorityscore=1501 malwarescore=0
 suspectscore=1 phishscore=0 bulkscore=0 spamscore=0 clxscore=1015
 lowpriorityscore=0 mlxscore=0 impostorscore=0 mlxlogscore=724 adultscore=0
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.0.1-1810050000
 definitions=main-1902040136
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

According to section "Checks on VMX Controls" in Intel SDM vol 3C, the
following check is performed on vmentry of L2 guests:

    On processors that support Intel 64 architecture, the IA32_SYSENTER_ESP
    field and the IA32_SYSENTER_EIP field must each contain a canonical
    address.

Signed-off-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
Reviewed-by: Mihai Carabas <mihai.carabas@oracle.com>
---
 arch/x86/kvm/vmx/nested.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/arch/x86/kvm/vmx/nested.c b/arch/x86/kvm/vmx/nested.c
index 7bb5e565f3fa..65b29726ad7a 100644
--- a/arch/x86/kvm/vmx/nested.c
+++ b/arch/x86/kvm/vmx/nested.c
@@ -2595,6 +2595,15 @@ static int nested_check_host_control_regs(struct kvm_vcpu *vcpu,
 	    !nested_host_cr4_valid(vcpu, vmcs12->host_cr4) ||
 	    !nested_cr3_valid(vcpu, vmcs12->host_cr3))
 		return -EINVAL;
+
+#ifdef CONFIG_X86_64
+	if (is_noncanonical_address(vmcs_readl(HOST_IA32_SYSENTER_ESP),
+						vcpu) ||
+	    is_noncanonical_address(vmcs_readl(HOST_IA32_SYSENTER_EIP),
+						vcpu))
+		return -EINVAL;
+#endif
+
 	/*
 	 * If the load IA32_EFER VM-exit control is 1, bits reserved in the
 	 * IA32_EFER MSR must be 0 in the field for that register. In addition,

From patchwork Mon Feb  4 17:16:50 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Krish Sadhukhan <krish.sadhukhan@oracle.com>
X-Patchwork-Id: 10796205
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id F0BFC1805
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  4 Feb 2019 17:41:57 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id E11E42BED4
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  4 Feb 2019 17:41:57 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id DFB342BEF6; Mon,  4 Feb 2019 17:41:57 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	UNPARSEABLE_RELAY autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 87D912BED4
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  4 Feb 2019 17:41:57 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728169AbfBDRl4 (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 4 Feb 2019 12:41:56 -0500
Received: from userp2130.oracle.com ([156.151.31.86]:57462 "EHLO
        userp2130.oracle.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727691AbfBDRl4 (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 4 Feb 2019 12:41:56 -0500
Received: from pps.filterd (userp2130.oracle.com [127.0.0.1])
        by userp2130.oracle.com (8.16.0.27/8.16.0.27) with SMTP id
 x14HXpVM019924;
        Mon, 4 Feb 2019 17:41:50 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : in-reply-to : references; s=corp-2018-07-02;
 bh=36eabbE1r7o21vNqo7N4v0J4yaIDOi84XARIn10cPpU=;
 b=owgoTFMYCvVPl2ssM1FDMu98zfyRBC9uCCmwMOFvGP2CQZKMdrBZjh7UpzF9WovggtZ/
 EHvXj6f2qMFrI0Rw6NJkoYutQh7EGkPLy+q/y1r2eXwbmNq8nxHvOPvkSFVDaZOi42Q5
 Z1sxRylo/xf/OufSdlx6zxw85Te++UCDYPURu+aPUvA36kZBUkXbM3iTpP0VHDKEWQ9a
 bMqYfBnH2EYUtdxW6DErPul7U6IJs+Ejue9XfynqyEQvM7W/1n0WgPkoyplyOoNQtIHt
 Gl9ZBmzpeMxZiM198Q/r2naGtgY939/XIPA3hK2UP6usjtrM/xduMGp7gXWfeN3zr/Qh Fg==
Received: from aserv0022.oracle.com (aserv0022.oracle.com [141.146.126.234])
        by userp2130.oracle.com with ESMTP id 2qd9ar6m2u-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Mon, 04 Feb 2019 17:41:49 +0000
Received: from aserv0121.oracle.com (aserv0121.oracle.com [141.146.126.235])
        by aserv0022.oracle.com (8.14.4/8.14.4) with ESMTP id x14HfmtO026569
        (version=TLSv1/SSLv3 cipher=DHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Mon, 4 Feb 2019 17:41:48 GMT
Received: from abhmp0018.oracle.com (abhmp0018.oracle.com [141.146.116.24])
        by aserv0121.oracle.com (8.14.4/8.13.8) with ESMTP id x14Hfm39019908;
        Mon, 4 Feb 2019 17:41:48 GMT
Received: from ban25x6uut29.us.oracle.com (/10.153.73.29)
        by default (Oracle Beehive Gateway v4.0)
        with ESMTP ; Mon, 04 Feb 2019 17:41:48 +0000
From: Krish Sadhukhan <krish.sadhukhan@oracle.com>
To: kvm@vger.kernel.org
Cc: pbonzini@redhat.com, rkrcmar@redhat.com, jmattson@google.com
Subject: [PATCH 2/2][kvm-unit-test nVMX]: Test HOST_SYSENTER_ESP and
 HOST_SYSENTER_EIP fields on vmentry of L2 guests
Date: Mon,  4 Feb 2019 12:16:50 -0500
Message-Id: <20190204171650.6920-3-krish.sadhukhan@oracle.com>
X-Mailer: git-send-email 2.17.2
In-Reply-To: <20190204171650.6920-1-krish.sadhukhan@oracle.com>
References: <20190204171650.6920-1-krish.sadhukhan@oracle.com>
X-Proofpoint-Virus-Version: vendor=nai engine=5900 definitions=9157
 signatures=668682
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0
 priorityscore=1501 malwarescore=0
 suspectscore=1 phishscore=0 bulkscore=0 spamscore=0 clxscore=1015
 lowpriorityscore=0 mlxscore=0 impostorscore=0 mlxlogscore=883 adultscore=0
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.0.1-1810050000
 definitions=main-1902040136
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

According to section "Checks on VMX Controls" in Intel SDM vol 3C, the
following check is performed on vmentry of L2 guests:

    On processors that support Intel 64 architecture, the IA32_SYSENTER_ESP
    field and the IA32_SYSENTER_EIP field must each contain a canonical
    address.

Signed-off-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
Reviewed-by: Mihai Carabas <mihai.carabas@oracle.com>
---
 x86/vmx_tests.c | 39 +++++++++++++++++++++++++++++++++++++++
 1 file changed, 39 insertions(+)

diff --git a/x86/vmx_tests.c b/x86/vmx_tests.c
index 9a3cdee..2e45e02 100644
--- a/x86/vmx_tests.c
+++ b/x86/vmx_tests.c
@@ -4909,6 +4909,44 @@ static void test_vm_exit_ctls(void)
 	test_exit_msr_store();
 }
 
+/*
+ * On processors that support Intel 64 architecture, the IA32_SYSENTER_ESP
+ * field and the IA32_SYSENTER_EIP field must each contain a canonical
+ * address.
+ *
+ *  [Intel SDM]
+ */
+static void test_host_ctl_regs(void)
+{
+	if (!is_canonical(vmcs_read(HOST_SYSENTER_ESP))) {
+		report_prefix_pushf("HOST_SYSENTER_ESP non-canonical");
+		test_vmx_controls(false, false);
+		report_prefix_pop();
+	} else {
+		report_prefix_pushf("HOST_SYSENTER_ESP canonical");
+		test_vmx_controls(true, false);
+		report_prefix_pop();
+	}
+
+	if (!is_canonical(vmcs_read(HOST_SYSENTER_EIP))) {
+		report_prefix_pushf("HOST_SYSENTER_EIP non-canonical");
+		test_vmx_controls(false, false);
+		report_prefix_pop();
+	} else {
+		report_prefix_pushf("HOST_SYSENTER_EIP canonical");
+		test_vmx_controls(true, false);
+		report_prefix_pop();
+	}
+}
+
+/*
+ * Tests for host control registers and MSRs
+ */
+static void test_vm_host_ctl_regs_msrs(void)
+{
+	test_host_ctl_regs();
+}
+
 /*
  * Check that the virtual CPU checks all of the VMX controls as
  * documented in the Intel SDM.
@@ -4925,6 +4963,7 @@ static void vmx_controls_test(void)
 	test_vm_execution_ctls();
 	test_vm_exit_ctls();
 	test_vm_entry_ctls();
+	test_vm_host_ctl_regs_msrs();
 }
 
 static bool valid_vmcs_for_vmentry(void)
